agreement/node.go

```diff
@@ -0,0 +1,367 @@
+				persistence.NewMessageMeta(EL_AG_UNABLE_SYNC_NODE_WITH_EXCH, err.Error()),
+				persistence.EC_ERROR_NODE_SYNC,
+				exchange.GetOrg(w.GetExchangeId()),
+				exchange.GetId(w.GetExchangeId()),
+				w.devicePattern, "")
+			w.isOffline()
+			return 0
+		}
+	} else {
+		w.hznOffline = false
+	}
+
+	glog.V(5).Infof(logString(fmt.Sprintf("Done checking exchange node changes.")))
+
+	// check the pattern changes
+	w.checkNodePatternChanges(exchNode)
+
+	// now check the user input changes.
+	w.checkNodeUserInputChanges(pDevice)
+
+	// check the service configstate changes
+	w.checkServiceConfigStateChanges(exchNode)
+
+	return 0
+}
+
+// Check the node user input changes on the exchange and sync up with
+// the local copy. THe exchange is the master
+func (w *AgreementWorker) checkNodePatternChanges(exchDevice *exchange.Device) {
+	glog.V(5).Infof(logString(fmt.Sprintf("checking the node pattern changes.")))
+
+	if exchDevice == nil {
+		return
+	}
+
+	glog.V(5).Infof(logString(fmt.Sprintf("checking the node pattern devp=%v, exchp=%v", w.devicePattern, exchDevice.Pattern)))
+
+	if w.devicePattern != "" && exchDevice.Pattern != "" && w.devicePattern != exchDevice.Pattern {
+		saved_pattern, err := persistence.FindSavedNodeExchPattern(w.db)
+		if err != nil {
+			glog.Errorf(logString(fmt.Sprintf("Unable to retrieve the saved node exchange pattern from the local database. %v", err)))
+			eventlog.LogDatabaseEvent(w.db, persistence.SEVERITY_ERROR,
+				persistence.NewMessageMeta(EL_AG_UNABLE_READ_NODE_EXCH_PATTERN_FROM_DB, err.Error()),
+				persistence.EC_DATABASE_ERROR)
+			return
+		} else if saved_pattern != "" {
+			// will not handle it because the pattern may be changed by the shutdown process.
+			glog.Infof(logString(fmt.Sprintf("Node pattern changed to %v on the exchange, but will not handle it because there is already a saved exchange pattern %v in local database that needs to be handled.", exchDevice.Pattern, saved_pattern)))
+			return
+		}
+
+		// save the node change pattern into the local db
+		if err := persistence.SaveNodeExchPattern(w.db, exchDevice.Pattern); err != nil {
+			glog.Errorf(logString(fmt.Sprintf("Unable to save the new node exchange pattern %v to the local database. %v", exchDevice.Pattern, err)))
+			eventlog.LogDatabaseEvent(w.db, persistence.SEVERITY_ERROR,
+				persistence.NewMessageMeta(EL_AG_UNABLE_WRITE_NODE_EXCH_PATTERN_TO_DB, exchDevice.Pattern, err.Error()),
+				persistence.EC_DATABASE_ERROR)
+			return
+		} else {
+			glog.Infof(logString(fmt.Sprintf("Node pattern changed on the exchange from %v to %v. Will re-register the node.", w.devicePattern, exchDevice.Pattern)))
+			eventlog.LogDatabaseEvent(w.db, persistence.SEVERITY_INFO,
+				persistence.NewMessageMeta(EL_AG_NODE_PATTERN_CHANGED, w.devicePattern, exchDevice.Pattern),
+				persistence.EC_NODE_PATTERN_CHANGED)
+			w.Messages() <- events.NewNodePatternMessage(events.NODE_PATTERN_CHANGE_SHUTDOWN, exchDevice.Pattern)
```

Comment:
You can't send the shutdown event until you verify that the new pattern can run on this node, i.e. that all the services in the new pattern have been configured on the device. If the new pattern cant run here then we should log that in the error log and perhaps this is one of the errors we should surface to the exchange.

Reply:
Added code to do verification.